contracts/EuroToken.sol

```diff
@@ -0,0 +1,235 @@
+    // Modifiers
+    ////////////////////////
+
+    modifier onlyAllowedTransferFrom(address from) {
+        require(_allowedTransferFrom[from]);
+        _;
+    }
+
+    modifier onlyAllowedTransferTo(address to) {
+        require(_allowedTransferTo[to]);
+        _;
+    }
+
+    ////////////////////////
+    // Constructor
+    ////////////////////////
+
+    function EuroToken(IAccessPolicy accessPolicy)
+        AccessControlled(accessPolicy)
+        StandardToken()
+        TokenMetadata(NAME, DECIMALS, SYMBOL, "")
+        MigrationSource(accessPolicy, ROLE_EURT_DEPOSIT_MANAGER)
+        Reclaimable()
+    {
+    }
+
+    ////////////////////////
+    // Public functions
+    ////////////////////////
+
+    /// @notice deposit 'amount' of EUR-T to address 'to'
+    /// @dev address 'to' is whitelisted as recipient of future transfers
+    /// @dev deposit may happen only in case of succesful KYC of recipient and validation of banking data
+    /// @dev which in this implementation is an off-chain responsibility of EURT_DEPOSIT_MANAGER
+    function deposit(address to, uint256 amount)
+        public
+        payable
+        only(ROLE_EURT_DEPOSIT_MANAGER)
+        returns (bool)
+    {
+        // cannot receive ether
+        require(msg.value == 0);
+        _balances[to] = add(_balances[to], amount);
+        _totalSupply = add(_totalSupply, amount);
+        setAllowedTransferTo(to, true);
+        Deposit(to, amount);
+        return true;
+    }
+
+    /// @notice withdraws 'amount' of EUR-T by burning required amount and providing a proof of whithdrawal
+    /// @dev proof is provided in form of log entry on which EURT_DEPOSIT_MANAGER
+    /// @dev will act off-chain to return required Euro amount to EUR-T holder
+    function withdraw(uint256 amount)
+        public
+    {
+        require(_balances[msg.sender] >= amount);
+        _balances[msg.sender] -= amount;
+        _totalSupply -= amount;
+        Withdrawal(msg.sender, amount);
+    }
+
+    /// @notice enables or disables address to be receipient of EUR-T
+    function setAllowedTransferTo(address to, bool allowed)
+        public
+        only(ROLE_EURT_DEPOSIT_MANAGER)
+    {
+        _allowedTransferTo[to] = allowed;
+        AllowedToAddress(to, allowed);
+    }
+
+    /// @notice enables or disables address to be sender of EUR-T
+    function setAllowedTransferFrom(address from, bool allowed)
+        public
+        only(ROLE_EURT_DEPOSIT_MANAGER)
+    {
+        _allowedTransferFrom[from] = allowed;
+        AllowedFromAddress(from, allowed);
+    }
+
+    function allowedTransferTo(address to)
+        public
+        constant
+        returns (bool)
+    {
+        return _allowedTransferTo[to];
+    }
+
+    function allowedTransferFrom(address from)
+        public
+        constant
+        returns (bool)
+    {
+        return _allowedTransferFrom[from];
+    }
+
+    //
+    // Overrides ERC20 Interface to allow transfer from/to allowed addresses
```

Comment:
If we start from the `SnapshotToken`, there are internal callbacks `mOnTransfer` and `mOnApprove` that can be used. Makes it slightly simpler to implement this. (fewer functions to override, no risk that you forget one, no need to call the base class, etc.)